sql/src/main/java/io/crate/operation/collect/collectors/LuceneBatchIterator.java

```diff
@@ -0,0 +1,252 @@
+import io.crate.operation.InputRow;
+import io.crate.operation.reference.doc.lucene.CollectorContext;
+import io.crate.operation.reference.doc.lucene.LuceneCollectorExpression;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.*;
+import org.apache.lucene.util.Bits;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Predicate;
+
+public class LuceneBatchIterator implements BatchIterator {
+
+    private final IndexSearcher indexSearcher;
+    private final Query query;
+    private final CollectorContext collectorContext;
+    private final RamAccountingContext ramAccountingContext;
+    private final boolean doScores;
+    private final Predicate<Scorer> filter;
+    private final LuceneCollectorExpression[] expressions;
+    private final List<LeafReaderContext> leaves;
+    private final InputRow inputRow;
+
+    private DocReaderConsumer docReaderConsumer;
+    private Row currentRow;
+    private Iterator<LeafReaderContext> leavesIt;
+    private Weight weight;
+    private LeafReaderContext currentLeaf;
+    private Scorer currentScorer;
+    private DocIdSetIterator currentDocIdSetIt;
+    private boolean closed = false;
+
+    public LuceneBatchIterator(IndexSearcher indexSearcher,
+                               Query query,
+                               Float minScore,
+                               boolean doScores,
+                               CollectorContext collectorContext,
+                               RamAccountingContext ramAccountingContext,
+                               List<? extends Input<?>> inputs,
+                               Collection<? extends LuceneCollectorExpression<?>> expressions) {
+        this.indexSearcher = indexSearcher;
+        this.query = query;
+        this.doScores = doScores || minScore != null;
+        this.filter = createMinScorePredicate(minScore);
+        this.collectorContext = collectorContext;
+        this.ramAccountingContext = ramAccountingContext;
+        this.inputRow = new InputRow(inputs);
+        this.currentRow = inputRow;
+        this.expressions = expressions.toArray(new LuceneCollectorExpression[0]);
+        leaves = indexSearcher.getTopReaderContext().leaves();
+        leavesIt = leaves.iterator();
+    }
+
+
+    @Override
+    public void moveToStart() {
+        raiseIfClosed();
+        leavesIt = leaves.iterator();
+        currentRow = OFF_ROW;
+    }
+
+    @Override
+    public boolean moveNext() {
+        raiseIfClosed();
+        try {
+            return innerMoveNext();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private boolean innerMoveNext() throws IOException {
+        if (weight == null) {
+            weight = createWeight();
```

Comment:
Wanted to avoid I/O access in the constructor - I also considered using loadNextBatch for the weight creation and start the iterator in a unloaded state. But that added a bit of complexity due to the loading state management.

I'm fine with moving it into the constructor if you think it's okay to have I/O there.